File: PriorityQueueUsingLL.java
import java.util.* ; 
  
//Linked List |  peek() O(1)    push()  O(n)     pop() O(1)

class Solution 
{ 
    
    
// Node  
 static class Node {  
    int data;  
    
    // Lower values indicate higher priority  
    int priority;  
    
     Node next;  
    
} 
  
static Node node = new Node(); 
    
// Function to Create A New Node  
static Node newNode(int d, int p)  
{  
    Node temp = new Node();  
    temp.data = d;  
    temp.priority = p;  
    temp.next = null;  
    
    return temp;  
}  
    
// Return the value at head  
static int peek(Node head)  
{  
    return (head).data;  
}  
    
// Removes the element with the  
// highest priority form the list  
static Node pop(Node head)  
{  
    Node temp = head;  
    (head)  = (head).next;  
    return head; 
}    
    
// Function to push according to priority  
static Node push(Node head, int d, int p)  
{  
    Node start = (head);  
    
    // Create new Node  
    Node temp = newNode(d, p);  
    
    // Special Case: The head of list has lesser  
    // priority than new node. So insert new  
    // node before head node and change head node.  
    if ((head).priority > p) {  
    
        // Insert New Node before head  
        temp.next = head;  
        (head) = temp;  
    }  
    else {  
    
        // Traverse the list and find a  
        // position to insert new node  
        while (start.next != null &&  
               start.next.priority < p) {  
            start = start.next;  
        }  
    
        // Either at the ends of the list  
        // or at required position  
        temp.next = start.next;  
        start.next = temp;  
    }  
    return head; 
}  
    
// Function to check is list is empty  
static int isEmpty(Node head)  
{  
    return ((head) == null)?1:0;  
}  
    

public static void main(String args[]) 
{  
    // Create a Priority Queue  
    // 3 2 1 4   
    Node pq = newNode(2, 1);  
    pq =push(pq, 1, 2);  
    pq =push(pq, 4, 3);  
    pq =push(pq, 3, 0);  
    
    while (isEmpty(pq)==0) {  
        System.out.printf("%d ", peek(pq));  
        pq=pop(pq);  
    }  
    
}  
} 

File: priority_queue_using_LL.py
class Node:
  def __init__(self, data):
    self.data = data 
    self.next = None

class PriorityQueueUsingLinkedLists:
  def __init__(self):
    self.head = Node(None)

  def push(self, value):
    if self.head.data is None:
      self.head = Node(value)
      return
    else:
      prev = None
      curr = self.head
      # print("curr.data",curr.data)
      while curr and curr.data > value:
          prev = curr
          curr = curr.next 
      new_node = Node(value)
      if prev == None: # if the new node is greater than all nodes 
        new_node.next = self.head
        self.head = new_node
      else:
        temp = prev.next
        prev.next = new_node 
        new_node.next = temp

  def pop(self):
    if not self.head:
      return 
    curr_head = self.head
    temp = self.head.next 
    self.head.next = None
    self.head = temp
    return curr_head.data

  def isEmpty(self):
    pass

  def printLL(self):
    print("Printing the contents of the PQ")
    if self.head:
      curr = self.head
      while curr:
        print(curr.data)
        curr = curr.next

pq = PriorityQueueUsingLinkedLists()
pq.push(3)
pq.push(2)
pq.push(1)
pq.push(1)
pq.push(2)
pq.push(3)
pq.push(7)
pq.push(5)
pq.push(6)
pq.push(7)
print("Before popping")
pq.printLL()
pq.pop()
pq.pop()
print("After popping")
pq.printLL()
# pq.isEmpty()
